File: src/agent_cap.py
"""Workers agent capability tokens — attenuating authority without in-place escalation."""
from __future__ import annotations

import hashlib
import hmac
import json
import math
from dataclasses import dataclass
from enum import Enum
from typing import Iterable


TOKEN_VERSION = 2


def canonical_json(obj: object) -> str:
    return json.dumps(
        obj,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    )


def digest(obj: object) -> str:
    return hashlib.sha256(canonical_json(obj).encode("utf-8")).hexdigest()


def _token(name: str, value: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(name)
    return value


def _finite(name: str, value: float) -> float:
    if not isinstance(value, (int, float)) or isinstance(value, bool) or not math.isfinite(float(value)):
        raise ValueError(name)
    return float(value)


def _capabilities(values: Iterable[str]) -> frozenset[str]:
    try:
        caps = frozenset(values)
    except TypeError as exc:
        raise ValueError("capabilities") from exc
    if not caps:
        raise ValueError("capabilities")
    for cap in caps:
        _token("capability", cap)
    return caps


class CapVerdict(str, Enum):
    ALLOW = "ALLOW"
    REFUSE = "REFUSE"


@dataclass(frozen=True)
class CapToken:
    request_id: str
    capabilities: frozenset[str]
    not_after: float
    mac: str
    not_before: float = 0.0
    issuer_id: str = "local-authority"
    parent_fingerprint: str = ""
    delegation_depth: int = 0
    version: int = TOKEN_VERSION

    def __post_init__(self) -> None:
        _token("request_id", self.request_id)
        object.__setattr__(self, "capabilities", _capabilities(self.capabilities))
        nb = _finite("not_before", self.not_before)
        na = _finite("not_after", self.not_after)
        if na <= nb:
            raise ValueError("validity_window")
        object.__setattr__(self, "not_before", nb)
        object.__setattr__(self, "not_after", na)
        _token("issuer_id", self.issuer_id)
        if not isinstance(self.delegation_depth, int) or isinstance(self.delegation_depth, bool) or self.delegation_depth < 0:
            raise ValueError("delegation_depth")
        if self.parent_fingerprint:
            if len(self.parent_fingerprint) != 64 or any(ch not in "0123456789abcdef" for ch in self.parent_fingerprint):
                raise ValueError("parent_fingerprint")
        elif self.delegation_depth != 0:
            raise ValueError("delegation_without_parent")
        if not isinstance(self.version, int) or isinstance(self.version, bool) or self.version != TOKEN_VERSION:
            raise ValueError("token_version")
        if not isinstance(self.mac, str) or len(self.mac) != 64 or any(ch not in "0123456789abcdef" for ch in self.mac):
            raise ValueError("mac")

    def unsigned_body(self) -> dict[str, object]:
        return {
            "version": self.version,
            "request_id": self.request_id,
            "capabilities": sorted(self.capabilities),
            "not_before": self.not_before,
            "not_after": self.not_after,
            "issuer_id": self.issuer_id,
            "parent_fingerprint": self.parent_fingerprint,
            "delegation_depth": self.delegation_depth,
        }

    def fingerprint(self) -> str:
        return digest({**self.unsigned_body(), "mac": self.mac})


@dataclass(frozen=True)
class DelegationReceipt:
    parent_fingerprint: str
    child_fingerprint: str
    issuer_id: str
    child_request_id: str
    removed_capabilities: tuple[str, ...]
    parent_not_before: float
    child_not_before: float
    parent_not_after: float
    child_not_after: float
    parent_depth: int
    child_depth: int
    fingerprint: str


class CapMint:
    def __init__(self, secret: bytes, issuer_id: str = "local-authority"):
        if not isinstance(secret, bytes) or not secret:
            raise ValueError("secret")
        self._secret = secret
        self.issuer_id = _token("issuer_id", issuer_id)

    def _mac(self, body: dict[str, object]) -> str:
        return hmac.new(
            self._secret,
            canonical_json(body).encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()

    def _build(
        self,
        request_id: str,
        capabilities: Iterable[str],
        *,
        not_before: float,
        not_after: float,
        parent_fingerprint: str,
        delegation_depth: int,
    ) -> CapToken:
        rid = _token("request_id", request_id)
        caps = _capabilities(capabilities)
        nb = _finite("not_before", not_before)
        na = _finite("not_after", not_after)
        if na <= nb:
            raise ValueError("validity_window")
        body: dict[str, object] = {
            "version": TOKEN_VERSION,
            "request_id": rid,
            "capabilities": sorted(caps),
            "not_before": nb,
            "not_after": na,
            "issuer_id": self.issuer_id,
            "parent_fingerprint": parent_fingerprint,
            "delegation_depth": delegation_depth,
        }
        return CapToken(
            request_id=rid,
            capabilities=caps,
            not_after=na,
            mac=self._mac(body),
            not_before=nb,
            issuer_id=self.issuer_id,
            parent_fingerprint=parent_fingerprint,
            delegation_depth=delegation_depth,
        )

    def mint(
        self,
        request_id: str,
        capabilities: Iterable[str],
        not_after: float,
        not_before: float = 0.0,
    ) -> CapToken:
        return self._build(
            request_id,
            capabilities,
            not_before=not_before,
            not_after=not_after,
            parent_fingerprint="",
            delegation_depth=0,
        )

    def verify(self, token: CapToken) -> bool:
        if not isinstance(token, CapToken):
            return False
        if token.issuer_id != self.issuer_id:
            return False
        try:
            expected = self._mac(token.unsigned_body())
        except (TypeError, ValueError):
            return False
        return hmac.compare_digest(expected, token.mac)

    def issue_subcap(
        self,
        parent: CapToken,
        request_id: str,
        capabilities: Iterable[str],
        not_after: float,
        not_before: float | None = None,
    ) -> tuple[CapToken, DelegationReceipt]:
        if not self.verify(parent):
            raise PermissionError("PARENT_TOKEN_INVALID")
        rid = _token("request_id", request_id)
        if rid == parent.request_id:
            raise ValueError("CHILD_REQUEST_ID_MUST_DIFFER")
        child_caps = _capabilities(capabilities)
        if not child_caps.issubset(parent.capabilities):
            raise PermissionError("CAPABILITY_ESCALATION")
        child_nb = parent.not_before if not_before is None else _finite("not_before", not_before)
        child_na = _finite("not_after", not_after)
        if child_nb < parent.not_before or child_na > parent.not_after or child_na <= child_nb:
            raise PermissionError("TIME_ESCALATION")

        child = self._build(
            rid,
            child_caps,
            not_before=child_nb,
            not_after=child_na,
            parent_fingerprint=parent.fingerprint(),
            delegation_depth=parent.delegation_depth + 1,
        )
        receipt = self.delegation_receipt(parent, child)
        return child, receipt

    def verify_delegation(self, parent: CapToken, child: CapToken) -> bool:
        if not self.verify(parent) or not self.verify(child):
            return False
        if child.parent_fingerprint != parent.fingerprint():
            return False
        if child.issuer_id != parent.issuer_id:
            return False
        if child.delegation_depth != parent.delegation_depth + 1:
            return False
        if not child.capabilities.issubset(parent.capabilities):
            return False
        if child.not_before < parent.not_before or child.not_after > parent.not_after:
            return False
        return True

    def delegation_receipt(self, parent: CapToken, child: CapToken) -> DelegationReceipt:
        if not self.verify_delegation(parent, child):
            raise PermissionError("DELEGATION_INVALID")
        removed = tuple(sorted(parent.capabilities - child.capabilities))
        body = {
            "parent_fingerprint": parent.fingerprint(),
            "child_fingerprint": child.fingerprint(),
            "issuer_id": child.issuer_id,
            "child_request_id": child.request_id,
            "removed_capabilities": list(removed),
            "parent_not_before": parent.not_before,
            "child_not_before": child.not_before,
            "parent_not_after": parent.not_after,
            "child_not_after": child.not_after,
            "parent_depth": parent.delegation_depth,
            "child_depth": child.delegation_depth,
        }
        return DelegationReceipt(
            parent_fingerprint=body["parent_fingerprint"],
            child_fingerprint=body["child_fingerprint"],
            issuer_id=child.issuer_id,
            child_request_id=child.request_id,
            removed_capabilities=removed,
            parent_not_before=parent.not_before,
            child_not_before=child.not_before,
            parent_not_after=parent.not_after,
            child_not_after=child.not_after,
            parent_depth=parent.delegation_depth,
            child_depth=child.delegation_depth,
            fingerprint=digest(body),
        )


class AgentCapRuntime:
    def __init__(self, mint: CapMint):
        self._mint = mint

    def invoke(self, token: CapToken, capability: str, now: float) -> tuple[CapVerdict, str | None]:
        if not self._mint.verify(token):
            return CapVerdict.REFUSE, "BAD_MAC"
        try:
            current = _finite("now", now)
            requested = _token("capability", capability)
        except ValueError:
            return CapVerdict.REFUSE, "INVALID_INVOCATION"
        if current < token.not_before:
            return CapVerdict.REFUSE, "NOT_YET_VALID"
        if current > token.not_after:
            return CapVerdict.REFUSE, "EXPIRED"
        if requested not in token.capabilities:
            return CapVerdict.REFUSE, "CAPABILITY_NOT_GRANTED"
        return CapVerdict.ALLOW, None

    def escalate(self, token: CapToken, new_cap: str) -> None:
        """Explicitly unsupported — authority must be issued as a new attenuated subcap."""
        raise RuntimeError("ESCALATION_FORBIDDEN: use issue_subcap with a verified parent")

File: ts/agent_cap.mjs
import crypto from "node:crypto";

export const TOKEN_VERSION = 2;
const TIME_SCALE = 1_000_000;

function token(name, value) {
  if (typeof value !== "string" || value.trim() === "") throw new TypeError(name);
  return value;
}

function finite(name, value) {
  if (typeof value !== "number" || !Number.isFinite(value)) throw new TypeError(name);
  return value;
}

function timeUs(name, value) {
  const normalized = finite(name, value);
  if (normalized < 0) throw new TypeError(name);
  const micros = Math.round(normalized * TIME_SCALE);
  if (!Number.isSafeInteger(micros) || Math.abs(micros / TIME_SCALE - normalized) > 1e-9) throw new TypeError(`${name}_precision`);
  return micros;
}

function caps(values) {
  const unique = [...new Set([...values])].sort();
  if (!unique.length) throw new TypeError("capabilities");
  for (const cap of unique) token("capability", cap);
  return unique;
}

function canonical(value) {
  if (value === null || typeof value === "string" || typeof value === "boolean" || typeof value === "number") {
    if (typeof value === "number" && !Number.isFinite(value)) throw new TypeError("non_finite_json");
    return value;
  }
  if (Array.isArray(value)) return value.map(canonical);
  if (typeof value === "object") {
    const out = {};
    for (const key of Object.keys(value).sort()) out[key] = canonical(value[key]);
    return out;
  }
  throw new TypeError("unsupported_json");
}

export function canonicalJson(value) { return JSON.stringify(canonical(value)); }
export function digest(value) { return crypto.createHash("sha256").update(canonicalJson(value), "utf8").digest("hex"); }

function unsignedBody({ requestId, capabilities, notBefore, notAfter, issuerId, parentFingerprint, delegationDepth, version }) {
  return {
    version,
    request_id: requestId,
    capabilities: [...capabilities].sort(),
    not_before_us: timeUs("not_before", notBefore),
    not_after_us: timeUs("not_after", notAfter),
    issuer_id: issuerId,
    parent_fingerprint: parentFingerprint,
    delegation_depth: delegationDepth,
  };
}

function mac(secret, body) {
  if (typeof secret !== "string" || !secret.length) throw new TypeError("secret");
  return crypto.createHmac("sha256", secret).update(canonicalJson(body), "utf8").digest("hex");
}

function build(secret, requestId, capabilities, notAfter, { notBefore = 0, issuerId = "local-authority", parentFingerprint = "", delegationDepth = 0 } = {}) {
  const rid = token("request_id", requestId);
  const normalizedCaps = caps(capabilities);
  const nb = finite("not_before", notBefore);
  const na = finite("not_after", notAfter);
  timeUs("not_before", nb); timeUs("not_after", na);
  token("issuer_id", issuerId);
  if (na <= nb) throw new TypeError("validity_window");
  if (!Number.isInteger(delegationDepth) || delegationDepth < 0) throw new TypeError("delegation_depth");
  if (parentFingerprint && !/^[0-9a-f]{64}$/.test(parentFingerprint)) throw new TypeError("parent_fingerprint");
  if (!parentFingerprint && delegationDepth !== 0) throw new TypeError("delegation_without_parent");
  const body = unsignedBody({ requestId: rid, capabilities: normalizedCaps, notBefore: nb, notAfter: na, issuerId, parentFingerprint, delegationDepth, version: TOKEN_VERSION });
  return Object.freeze({ requestId: rid, capabilities: Object.freeze(normalizedCaps), notBefore: nb, notAfter: na, issuerId, parentFingerprint, delegationDepth, version: TOKEN_VERSION, mac: mac(secret, body) });
}

export function mint(secret, requestId, capabilities, notAfter, options = {}) { return build(secret, requestId, capabilities, notAfter, options); }
export function fingerprint(tokenValue) { return digest({ ...unsignedBody(tokenValue), mac: tokenValue.mac }); }

export function verify(secret, tokenValue, issuerId = "local-authority") {
  try {
    if (!tokenValue || tokenValue.version !== TOKEN_VERSION || tokenValue.issuerId !== issuerId) return false;
    if (!Array.isArray(tokenValue.capabilities) || !tokenValue.capabilities.length) return false;
    const normalizedCaps = caps(tokenValue.capabilities);
    if (canonicalJson(normalizedCaps) !== canonicalJson(tokenValue.capabilities)) return false;
    const body = unsignedBody({ ...tokenValue, capabilities: normalizedCaps });
    const expected = mac(secret, body);
    if (typeof tokenValue.mac !== "string" || !/^[0-9a-f]{64}$/.test(tokenValue.mac)) return false;
    return crypto.timingSafeEqual(Buffer.from(expected, "hex"), Buffer.from(tokenValue.mac, "hex"));
  } catch { return false; }
}

export function issueSubcap(secret, parent, childRequestId, capabilities, notAfter, options = {}) {
  const issuerId = options.issuerId ?? parent?.issuerId ?? "local-authority";
  if (!verify(secret, parent, issuerId)) throw new Error("PARENT_TOKEN_INVALID");
  const rid = token("request_id", childRequestId);
  if (rid === parent.requestId) throw new Error("CHILD_REQUEST_ID_MUST_DIFFER");
  const childCaps = caps(capabilities);
  if (!childCaps.every(capability => parent.capabilities.includes(capability))) throw new Error("CAPABILITY_ESCALATION");
  const childNotBefore = options.notBefore ?? parent.notBefore;
  timeUs("not_before", childNotBefore); timeUs("not_after", notAfter);
  if (childNotBefore < parent.notBefore || notAfter > parent.notAfter || notAfter <= childNotBefore) throw new Error("TIME_ESCALATION");
  const child = build(secret, rid, childCaps, notAfter, { notBefore: childNotBefore, issuerId, parentFingerprint: fingerprint(parent), delegationDepth: parent.delegationDepth + 1 });
  return { child, receipt: delegationReceipt(secret, parent, child) };
}

export function verifyDelegation(secret, parent, child) {
  const issuerId = parent?.issuerId ?? "local-authority";
  if (!verify(secret, parent, issuerId) || !verify(secret, child, issuerId)) return false;
  return child.parentFingerprint === fingerprint(parent)
    && child.issuerId === parent.issuerId
    && child.delegationDepth === parent.delegationDepth + 1
    && child.capabilities.every(capability => parent.capabilities.includes(capability))
    && child.notBefore >= parent.notBefore
    && child.notAfter <= parent.notAfter;
}

export function delegationReceipt(secret, parent, child) {
  if (!verifyDelegation(secret, parent, child)) throw new Error("DELEGATION_INVALID");
  const removedCapabilities = parent.capabilities.filter(capability => !child.capabilities.includes(capability)).sort();
  const body = {
    parent_fingerprint: fingerprint(parent), child_fingerprint: fingerprint(child), issuer_id: child.issuerId,
    child_request_id: child.requestId, removed_capabilities: removedCapabilities,
    parent_not_before_us: timeUs("parent_not_before", parent.notBefore), child_not_before_us: timeUs("child_not_before", child.notBefore),
    parent_not_after_us: timeUs("parent_not_after", parent.notAfter), child_not_after_us: timeUs("child_not_after", child.notAfter),
    parent_depth: parent.delegationDepth, child_depth: child.delegationDepth,
  };
  return Object.freeze({ parentFingerprint: body.parent_fingerprint, childFingerprint: body.child_fingerprint, issuerId: child.issuerId, childRequestId: child.requestId, removedCapabilities: Object.freeze(removedCapabilities), parentNotBefore: parent.notBefore, childNotBefore: child.notBefore, parentNotAfter: parent.notAfter, childNotAfter: child.notAfter, parentDepth: parent.delegationDepth, childDepth: child.delegationDepth, fingerprint: digest(body) });
}

export function invoke(secret, tokenValue, capability, now) {
  if (!verify(secret, tokenValue, tokenValue?.issuerId ?? "local-authority")) return { status: "REFUSE", reason: "BAD_MAC" };
  if (typeof capability !== "string" || capability.trim() === "" || typeof now !== "number" || !Number.isFinite(now)) return { status: "REFUSE", reason: "INVALID_INVOCATION" };
  if (now < tokenValue.notBefore) return { status: "REFUSE", reason: "NOT_YET_VALID" };
  if (now > tokenValue.notAfter) return { status: "REFUSE", reason: "EXPIRED" };
  if (!tokenValue.capabilities.includes(capability)) return { status: "REFUSE", reason: "CAPABILITY_NOT_GRANTED" };
  return { status: "ALLOW", reason: null };
}

export function escalate() { throw new Error("ESCALATION_FORBIDDEN: use issueSubcap with a verified parent"); }
